### ISSUE TITLE
Wrong parameter order in parameter covariance matrix?

### ISSUE THREAD

**Issue by dlakaplan** [nanograv/PINT#1043]
I am trying to use the parameter covariance matrix in a manner similar to `https://github.com/nanograv/PINT/blob/master/src/pint/random_models.py`

However, it seems that the parameter order in the covariance matrix output is not correct.

I tried to make two fake data-sets, following https://nanograv-pint.readthedocs.io/en/latest/examples/build_model_from_scratch.html

One has `RAJ`, `DECJ`, `F0`, `F1` free, the other only has `F0`, `F1` free.  Otherwise they are the same.

I printed out the covariance matrices after fitting, and checked the uncertainties computed from the covariance matrices  against those given by standard methods.

RAJ/DECJ unfrozen:
```
Parameter covariance matrix:
               RAJ       DECJ        F0         F1     
    RAJ      1.144e-27
   DECJ     -5.078e-36  4.702e-44
    F0       7.092e-25 -2.822e-33  1.499e-18
    F1       2.799e-22 -2.658e-30  1.286e-16  9.375e-14

Raw uncertainties (sqrt(diag(cov)):  [3.38253001e-14 2.16840735e-22 1.22444454e-09 3.06184736e-07]
Covariance matrix factors:  [1.00000000e+06 4.45790202e+12 8.84356289e+20 8.69554315e+07
 3.48003779e+05]
uncertainties:  odict_values([4.4080003405714652774e-06, 0.0011022650481010516, 3.3825300106387037565e-14, 2.1684073457865876917e-22])
```
RAJ/DECJ frozen:
```
Parameter covariance matrix:
               F0         F1     
    F0       1.143e-27
    F1      -5.070e-36  4.694e-44

Raw uncertainties (sqrt(diag(cov)):  [3.38117652e-14 2.16665758e-22]
Covariance matrix factors:  [1.00000000e+06 4.45790202e+12 8.84356289e+20]
```
In the second instance, the `F0` uncertainty is 3.38e-14.  This is true using the covariance matrix and the parameter interface.  But in the first instance the order is wrong, so it says that the uncertainty on `RAJ` is 3.38e-14 based on the covariance matrix (it should be 4.4e-6).  So the order of parameters in the covariance matrix appears incorrect.

The underlying fitting seems fine, so this is probably some issue in the order of the parameter dictionary vs. that in the matrix.

**Comment by dlakaplan** [comment]
Testing code:
```python
import astropy.units as u
import matplotlib.pyplot as plt
import numpy as np
import copy

import pint.fitter
from pint.models import get_model_and_toas
from pint.residuals import Residuals
from pint.toa import get_TOAs, make_fake_toas
from pint.phase import Phase

from pint.models import (
    parameter as p,
)  # We would like to add parameters to the model, so we need parameter module.
from pint.models.timing_model import (
    TimingModel,
    Component,
)  # Interface for timing model
import pint
from astropy.time import Time


# Make two timing models that are identical except for the free parameters
# largely taken from https://nanograv-pint.readthedocs.io/en/latest/examples/build_model_from_scratch.html
all_components = Component.component_types
selected_components = ["AbsPhase", "AstrometryEquatorial", "Spindown","SolarSystemShapiro"]
component_instances = []

# Initiate the component instances
for cp_name in selected_components:
    component_class = all_components[cp_name]  # Get the component class
    component_instance = component_class()  # Instantiate a component object
    component_instances.append(component_instance)
tm1 = TimingModel("TEST", component_instances)
tm2 = TimingModel("TEST", copy.deepcopy(component_instances))

# For `tm1`, `RAJ`, `DECJ`, `F0`, and `F1` are free.  For `tm2` keep `RAJ`, `DECJ` frozen
params = {
    "PSR": ("1748-2021E",),
    "RAJ": ("17:48:52.75", 1, 0.05 * pint.hourangle_second),
    "DECJ": ("-20:21:29.0", 1, 0.4 * u.arcsec),
    "F0": (61.485476554 * u.Hz, 1, 5e-10 * u.Hz),
    "F1": (-1.181e-15 * u.Hz / u.s, 1, 1e-18 * u.Hz / u.s),
    "PEPOCH": (Time(53750.000000, format="mjd", scale="tdb"),),
    "POSEPOCH": (Time(53750.000000, format="mjd", scale="tdb"),),
    "TZRMJD": (Time(53801.38605120074849, format="mjd", scale="tdb"),),
    "TZRFRQ": (1949.609 * u.MHz,),
    "TZRSITE": (1,),
}

# Assign the parameters
for name, info in params.items():
    par = getattr(tm1, name)  # Get parameter object from name
    par.quantity = info[0]  # set parameter value
    if len(info) > 1:
        if info[1] == 1:
            par.frozen = False  # Frozen means not fit.
        par.uncertainty = info[2]
for name, info in params.items():
    par = getattr(tm2, name)  # Get parameter object from name
    par.quantity = info[0]  # set parameter value
    if len(info) > 1:
        if info[1] == 1:
            par.frozen = False  # Frozen means not fit.
        par.uncertainty = info[2]
tm2.RAJ.frozen = True
tm2.DECJ.frozen = True

tm1.validate()
tm2.validate()
print('tm1: RAJ/DECJ unfrozen:')
print(tm1)
print('tm2: RAJ/DECJ frozen:')
print(tm2)


# make some TOAs
t1 = pint.toa.make_fake_toas(50000,60000,100,model=tm1)
t2 = pint.toa.make_fake_toas(50000,60000,100,model=tm2)

f1 = pint.fitter.WLSFitter(t1, tm1)
f1.fit_toas()
print('tm1: RAJ/DECJ unfrozen')
f1.print_summary()
print('tm2: RAJ/DECJ frozen')
f2 = pint.fitter.WLSFitter(t2, tm2)
f2.fit_toas()
f2.print_summary()

print('tm1: RAJ/DECJ unfrozen')
corm1 = f1.get_correlation_matrix(pretty_print=True)
cov1 = f1.get_covariance_matrix(pretty_print=True)
print('Raw uncertainties (sqrt(diag(cov)): ',np.sqrt(np.diag(cov1)))
print('Covariance matrix factors: ',f1.fac)
print('tm2: RAJ/DECJ frozen')
corm2 = f2.get_correlation_matrix(pretty_print=True)
cov2 = f2.get_covariance_matrix(pretty_print=True)
print('Raw uncertainties (sqrt(diag(cov)): ',np.sqrt(np.diag(cov2)))
print('Covariance matrix factors: ',f2.fac)

# get the parameter values and uncertainties from each model
# this is taken from `random_models.py`
params1 = f1.model.get_params_dict("free", "value")
params_u1 = f1.model.get_params_dict("free", "uncertainty")

mean_vector1 = params1.values()
uncertainty_vector1 = params_u1.values()
# remove the first column and row (absolute phase)
cov_matrix1 = (((f1.covariance_matrix[1:]).T)[1:]).T
fac1 = f1.fac[1:]

params2 = f2.model.get_params_dict("free", "value")
params_u2 = f2.model.get_params_dict("free", "uncertainty")

mean_vector2 = params2.values()
uncertainty_vector2 = params_u2.values()
# remove the first column and row (absolute phase)
cov_matrix2 = (((f2.covariance_matrix[1:]).T)[1:]).T
fac2 = f2.fac[1:]

print('\nThe values above should agree with the values below:')
print('tm1: RAJ/DECJ unfrozen')
print('mean values: ',mean_vector1)
print('uncertainties: ',uncertainty_vector1)
print('tm2: RAJ/DECJ frozen')
print('mean values: ',mean_vector2)
print('uncertainties: ',uncertainty_vector2)

```

**Comment by scottransom** [comment]
This is probably why the random_models() in pintk are not working correctly, as well.  I wonder if this is related to the "hidden" phase parameter?

**Comment by dlakaplan** [comment]
I don't think that (the phase parameter) is the issue, as that is explicitly dealt with. You can see that the blocks of `RAJ`/`DECJ` and `F0`/`F1` swap where they should be - it's not an off-by-one error

**Comment by dlakaplan** [comment]
(and yes, using the covariance matrix in this way messes up the routine I was trying to write which was similar to `random_models()`, but I didn't realize that wasn't working either)

**Comment by scottransom** [comment]
Ah, right.  Then the correct way to fix this is probably the same as the discussion that we had about the Design Matrix parameter mapping (and #1033) in last week's telecon.

**Comment by dlakaplan** [comment]
I'm working on this and making OK progress.  I have a few proposed UI changes to the `PintMatrix` class, but nothing major.

However, I'm wondering about terminology.  As far as I can tell we use `covariance_matrix` in different places for both the covariance of the TOAs and the parameters.  They have different needs/uses/sizes, but similar names. This can get confusing.  Would there be objections to at least making the latter (which is the one I am working on) `parameter_covariance_matrix` (and `parameter_correlation_matrix`)?

**Comment by scottransom** [comment]
@dlakaplan I think that is a good idea.  And @clp3ef and I noticed that when we were making the random models as well.

**Comment by dlakaplan** [comment]
A related question: the parameters, units, and matrix values are often re-defined on every iteration of a `Fitter` or `ModelState`.  For instance:
```
class WLSState(ModelState):
    def __init__(self, fitter, model, threshold=None):
        super().__init__(fitter, model)
        self.threshold = threshold

    @cached_property
    def step(self):
        # Define the linear system
        M, params, units = self.model.designmatrix(
            toas=self.fitter.toas, incfrozen=False, incoffset=True
        )
```
The `params` and `units` are then used to define the labels for the `CovarianceMatrix`.  My proposal is that we add attributes for `params`, `units`, and `parameter_labels` so that they will persist between iterations. I don't know that this would be a huge speed-up (I wouldn't remove the multiple calls to `self.model.designmatrix`) and they would need to be refreshed when parameters are added/deleted, but it might help organizationally.

**Comment by dlakaplan** [comment]
Also, consider deprecating `utils/show_param_cov_matrix` in favor of `CovarianceMatrix.prettyprint`

**Comment by dlakaplan** [comment]
I think I mostly have an implementation that works, including using `CovarianceMatrix` for most of the fitters (I am unsure of some of them, but `WLSFitter` and `GLSFitter` are OK).  This passes all of the tests except for blackening (can't figure out why).

I also made my own random model routine.  Usage is like:
```
    f.fit_toas()

    # this contains TOAs up through 54200
    # make new ones starting there
    tnew = toa.make_fake_toas(54200, 59000, 59000 - 54200, f.model)
    dphase, mrand = utils.calculate_random_models(f, tnew, Nmodels=100)
...
```
so this computes the phase for each model relative to the input, and returns all of the random models.  You can specify that only a subset of the parameters should be varied (e.g., `params=["F0"]`).  I think the results make sense:

![random_models](https://user-images.githubusercontent.com/5092062/118314695-18ec7b00-b4ba-11eb-8ec3-240613d36e3e.png)

The analytic phase uncertainty is just `~sqrt((dF0 * dt)**2 + (0.5 * dF1 * dt**2)**2)` (uncertainties from `F0` and `F1` added in quadrature), which agrees with the full version.   If I only vary `F0` then it's just linear in time.  There is also a test routine for this.

**Comment by dlakaplan** [comment]
(sorry - in the attachment the legend should say `F0 free`)

**Comment by dlakaplan** [comment]
I cannot get this to pass all tests. `black` fails, even though when I run it locally it doesn't need any changes.  Are there version or configuration issues I should be aware of?

**Comment by dlakaplan** [comment]
Maybe a question for @aarchiba : for some of the downhill fitters, is there a way to move some of the calculations of the covariance matrix out of the iteration step and store the info?  or maybe it doesn't make a difference and would get complicated as the number of parameters changes.

**Comment by aarchiba** [comment]
It certainly would be possible to move the calculations in the downhill fitters around, and it might accelerate things. The way the downhill fitters are implemented - linear algebra living in an object with cached values - might make this easier. Preserving matrices between runs would be complicated, because we receive no notification of what users do to the model (freezing parameters, say) before calling `fit_toas` again, but some things are recalculated even within a fitting run.

But before putting too much sweat into that, do some profiling. I found that for NANOGrav and for simple test workloads, the bulk of the fitter's time was spent having the model compute residuals and derivatives, and the linear algebra calculations were relatively cheap. (There is much room for optimization there - for example, the solar wind DM calculation uses the position, which is computed taking into account the proper motion and parallax for each TOA; and I believe that the solar system Roemer delay also uses these positions, also calculated afresh for each TOA; and all the proper motions use Astropy functions that handle full three-dimensional motion with a bogus distance of 1 kpc for all sources.)

**Comment by aarchiba** [comment]
I should check the code to be sure, but I initially computed the parameter covariance matrix only at the end of the fitting process (in a lazy property, at need). If it's being computed at each iteration step, it's probably because I found it useful, perhaps to determine a good stopping criterion? (No parameter changed by more than 0.01 sigma, perhaps?)

**Comment by aarchiba** [comment]
Is this closed by #1047?
